### DISCUSSION TITLE
Need non interactive option for configuring SSO on ssh keys and authorizing it for specific organizations

### DISCUSSION THREAD

**Discussion by aubreyyan** [community/community#89910] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

I am referring to this process https://docs.github.com/en/enterprise-cloud@latest/authentication/authenticating-with-saml-single-sign-on/authorizing-an-ssh-key-for-use-with-saml-single-sign-on#authorizing-an-ssh-key

There should be a way to do this through the gh cli.

Something like `gh ssh-key --configure-sso my-org authorize/deauthorize`

**Comment by sarahsturgeon** [comment]
Based on the way the URLs are generated for the process, it seems like it's intentionally made to be only interactive.
Not quite sure why _this_ would be the only part of the process that we can't automate or make easier, but, yeah.

---


A little bit of my findings while tinkering 

> [!NOTE]
> TL;DR it's a signed token and my findings are probably completely useless

The URL structure for configuring SSO on a key -> ORG _(that is, the link you get taken to when you click Configure SSO -> Click on an org)_ is:
```
https://github.com/orgs/<Org Name>/sso?authorization_request=<Long Key>&return_to=https%3A%2F%2Fgithub.com%2Fsettings%2Fkeys
```

The `authorization_request` param is a base-32-encoded binary blob with no padding.

Decoded, it has two parts:

**18-byte header:**
```
Bytes 0-3:    00 79 19 b7  — fixed magic bytes, identical across all keys, sessions, and reloads
Bytes 4-13:   (10 bytes)   — server-generated(?) nonce, changes every reload, differs between keys on same page
Byte 14:      69           — fixed across all observed samples
Byte 15:      a7 → a8      — slow-changing (noticed a change across days), stable across reloads and keys within a session.
                              only two values observed across two sessions. cause unknown — could be
                              daily rotation, server version, account state, etc.
Bytes 16-17:  (2 bytes)    — server-generated(?) nonce, changes every reload
```

**MessagePack payload (bytes 18+ 😏):**
```
{
    "organization_id": uint32,       # fetchable via GET /orgs/{name}
    "credential_id":   uint32,       # SSH key ID, fetchable via GET /user/keys
    "credential_type": "PublicKey"   # always this string for SSH keys
}
```

The 12 nonce bytes (`4-13`, `16-17`) and one slow-moving(?) byte (`15`) seem server-generated and can't be reproduced. Zeroing them produces a URL that completes the SSO auth flow for the org, but not the "authorize this key for this org" flow 🤷‍♀️ 

I could have tried across multiple accounts, browsers, and days to pin down more of those unknown bytes but ultimately it doesn't matter. It's almost certain that the token is intentionally signed to prevent forgery.
